contracts/src/Gateway.sol

```diff
@@ -108,6 +108,24 @@ contract Gateway is IGateway, IInitializable, IUpgradable {
         _;
     }
 
+    modifier nonreentrant() {
+        assembly {
+            // Check if flag is set and if true revert because it means the function is currently executing.
+            if tload(0) { revert(0, 0) }
```

Comment:
Would it be better to use a custom slot like
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/4c3ef87cf57b448a0b5fc68b8ce6604a31b60814/contracts/utils/ReentrancyGuardTransient.sol#L19 does?

Reply:
We can use that pattern. But in the V2 code we use slot 0? Is this an issue?

Reply:
It should be fine because all our variables except for DISPATCH_OVERHEAD_GAS are constant or immutable. DISPATCH_OVERHEAD_GAS is neither constant or immutable but we never write to it, so storage slot 0 is never used in Gateway. 
Just to be safe I made DISPATCH_OVERHEAD_GAS constant in:
https://github.com/Snowfork/snowbridge/pull/1358/commits/61481164a24ff79947decbf794b636b49f1fddab
And added a storage slot for the reentrancy guard so that we do not mistakenly re-use that slot 0 in future by adding variables to the mutable Gateway.
https://github.com/Snowfork/snowbridge/pull/1358/commits/266a35025b2e609af67fdcf81e01c15df43fce6c

Reply:
Transient storage lives in a separate data location from normal storage, at least that's what the internet says. Can check this with a test.

I generally don't like the way OpenZeppelin code is implemented, its verbose.